numpy/core/tests/test_numeric.py

```diff
@@ -2148,5 +2148,80 @@ def test_outer_out_param():
     assert_equal(res1, out1)
     assert_equal(np.outer(arr2, arr3, out2), out2)
 
+class TestRequire(object):
+    flag_names = ['C', 'C_CONTIGUOUS', 'CONTIGUOUS',
+                  'F', 'F_CONTIGUOUS',
+                  'A', 'ALIGNED',
+                  'W', 'WRITEABLE',
+                  'O', 'OWNDATA']
+
+    def generate_all_false(self, dtype):
+        arr = np.zeros((2, 2), [('junk', 'i1'), ('a', dtype)])
+        arr.setflags(write=False)
+        a = arr['a']
+        assert_(not a.flags['C'])
+        assert_(not a.flags['F'])
+        assert_(not a.flags['O'])
+        assert_(not a.flags['W'])
+        assert_(not a.flags['A'])
+        return a
+
+    def set_and_check_flag_no_dtype(self, flag, arr):
+        b = np.require(arr, None, [flag])
+        assert_(b.flags[flag])
+        # call in this way require should reserve the dtype
+        assert_(b.dtype == arr.dtype)
+
+        # a further call to np.require ought to return the same array
+        # unless OWNDATA is specified.
+        c = np.require(b, None, [flag])
+        if flag[0] != 'O':
+            assert_(c is b)
+        else:
+            assert_(c.flags[flag])
+
+    def test_require_each(self):
+       for dtype in ['f8', 'i4']:
+            for flag in self.flag_names:
+                a = self.generate_all_false(dtype)
+                yield self.set_and_check_flag_no_dtype, flag, a
+    
+    def test_unknown_requirement(self):
+        a = self.generate_all_false('f8')
+        assert_raises(KeyError, np.require, a, None, 'Q')
+
+    def test_non_array_input(self):
+        a = np.require([1, 2, 3, 4], 'i4', ['C', 'A', 'O'])
+        assert_(a.flags['O'])
+        assert_(a.flags['C'])
+        assert_(a.flags['A'])
+        assert_(a.dtype == 'i4')
+        assert_equal(a, [1, 2, 3, 4])
+
+    def test_C_and_F_simul(self):
+        # 2d arrays can't be both C and F simultaneously so this should raise
+        # but it doesn't as of numpy 1.9.1 and always returns a 'F' array.  So
+        # let's not break that (I guess?)
```

Comment:
I'd be tempted to raise and error in that case, as it is likely to be a programming mixup. Easy to check with the simplified set
```
if requirements.superset(['F', 'C']):
    raise ValueError(...)
```